File: backend/app/services/rag_retrieval.py
"""RAG retrieval — port of services/ragRetrieval.ts."""
from __future__ import annotations

import json
from pathlib import Path
from typing import List

from app.models import ProjectInput, RetrievedChunk, RegulatoryStatus

TOP_K = 5
MAX_CONTEXT_CHARS = 6000 * 4  # ~4 chars per token

_SOURCES_PATH = Path(__file__).parent / "knowledge" / "regulatory_sources.json"


def _build_queries(inp: ProjectInput) -> list[str]:
    queries = [
        "PV-Anlage Registrierung MaStR Pflichten Inbetriebnahme",
        "Netzanmeldung Netzbetreiber Einspeisevergütung EEG",
        "Steuerbefreiung Photovoltaik Einfamilienhaus §3 Nr.72 EStG",
    ]
    if inp.planWallbox or inp.hasEV:
        queries.append("§14a EnWG Wallbox Steuerbarkeit Netzentgelt")
    if inp.planStorage:
        queries.append("Batteriespeicher KfW Förderung Heimspeicher")
    if inp.planHeatPump:
        queries.append("Wärmepumpe §14a EnWG steuerbare Verbrauchseinrichtung")
    if inp.existingPvKwp:
        queries.append("Bestandsanlage EEG Vergütung Repowering Erweiterung")
    return queries


# ── Levenshtein / MMR helpers ────────────────────────────────────────────────

def _edit_dist(a: str, b: str) -> int:
    la, lb = len(a), len(b)
    dp = list(range(lb + 1))
    for i in range(1, la + 1):
        prev = dp[:]
        dp[0] = i
        for j in range(1, lb + 1):
            if a[i - 1] == b[j - 1]:
                dp[j] = prev[j - 1]
            else:
                dp[j] = 1 + min(prev[j], dp[j - 1], prev[j - 1])
    return dp[lb]


def _levenshtein_similarity(a: str, b: str) -> float:
    longer, shorter = (a, b) if len(a) >= len(b) else (b, a)
    if not longer:
        return 1.0
    return (len(longer) - _edit_dist(longer, shorter)) / len(longer)


def _mmr_deduplicate(chunks: list[RetrievedChunk]) -> list[RetrievedChunk]:
    selected: list[RetrievedChunk] = []
    seen: set[str] = set()
    total_chars = 0

    for chunk in sorted(chunks, key=lambda c: c.score, reverse=True):
        if chunk.id in seen:
            continue
        is_dup = any(
            s.source == chunk.source and _levenshtein_similarity(s.text, chunk.text) > 0.8
            for s in selected
        )
        if is_dup:
            continue
        total_chars += len(chunk.text)
        if total_chars > MAX_CONTEXT_CHARS:
            break
        seen.add(chunk.id)
        selected.append(chunk)

    return selected


# ── Static fallback ──────────────────────────────────────────────────────────

def _static_fallback(tag_filter: list[str]) -> list[RetrievedChunk]:
    sources = json.loads(_SOURCES_PATH.read_text(encoding="utf-8"))
    results = []
    for i, doc in enumerate(sources):
        if any(t in doc["tags"] for t in tag_filter):
            results.append(
                RetrievedChunk(
                    id=doc["id"],
                    text=doc["chunkText"],
                    source=doc["source"],
                    title=doc["title"],
                    status=RegulatoryStatus(doc["status"]),
                    validFrom=doc["validFrom"],
                    validUntil=doc.get("validUntil"),
                    tags=doc["tags"],
                    score=1.0 - i * 0.05,
                )
            )
    return results


# ── Main entry point ─────────────────────────────────────────────────────────

async def retrieve_relevant_chunks(inp: ProjectInput) -> List[RetrievedChunk]:
    queries = _build_queries(inp)
    tag_filter = ["pv"]
    if inp.planWallbox or inp.hasEV:
        tag_filter.append("wallbox")
    if inp.planStorage:
        tag_filter.append("storage")
    if inp.planHeatPump:
        tag_filter.append("heat_pump")

    all_chunks: list[RetrievedChunk] = []

    try:
        from app.services.knowledge.vector_db import get_collection

        collection = await get_collection()

        for query in queries:
            where = {"tags": {"$in": tag_filter}} if len(tag_filter) > 1 else None
            results = await collection.query(
                query_texts=[query],
                n_results=TOP_K,
                where=where,
            )

            ids = results["ids"][0] if results["ids"] else []
            distances = results["distances"][0] if results.get("distances") else []
            metadatas = results["metadatas"][0] if results.get("metadatas") else []
            documents = results["documents"][0] if results.get("documents") else []

            for idx, chunk_id in enumerate(ids):
                meta = metadatas[idx] or {}
                tags_raw = meta.get("tags", "")
                all_chunks.append(
                    RetrievedChunk(
                        id=str(chunk_id),
                        text=str(documents[idx] or ""),
                        source=str(meta.get("source", "")),
                        title=str(meta.get("title", "")),
                        status=RegulatoryStatus(meta.get("status", "unclear")),
                        validFrom=str(meta.get("validFrom", "")),
                        validUntil=meta.get("validUntil") or None,
                        tags=tags_raw.split(",") if tags_raw else [],
                        score=1.0 - float(distances[idx] if idx < len(distances) else 1.0),
                    )
                )
    except Exception:
        # Vector DB not available — fall back to static source metadata
        import warnings
        warnings.warn("Vector DB unavailable, falling back to static sources")
        return _static_fallback(tag_filter)

    return _mmr_deduplicate(all_chunks)

File: backend/src/services/ragRetrieval.ts
import type { ProjectInput, RetrievedChunk, RegulatoryDocument } from '../types';
import { getCollection } from '../knowledge/vectorDb';
import regulatorySources from '../knowledge/regulatorySources.json';

const TOP_K = 5;
const MAX_CONTEXT_TOKENS = 6000; // approx chars

function buildQueries(input: ProjectInput): string[] {
  const queries: string[] = [
    'PV-Anlage Registrierung MaStR Pflichten Inbetriebnahme',
    'Netzanmeldung Netzbetreiber Einspeisevergütung EEG',
    'Steuerbefreiung Photovoltaik Einfamilienhaus §3 Nr.72 EStG',
  ];

  if (input.planWallbox || input.hasEV) {
    queries.push('§14a EnWG Wallbox Steuerbarkeit Netzentgelt');
  }
  if (input.planStorage) {
    queries.push('Batteriespeicher KfW Förderung Heimspeicher');
  }
  if (input.planHeatPump) {
    queries.push('Wärmepumpe §14a EnWG steuerbare Verbrauchseinrichtung');
  }
  if (input.existingPvKwp) {
    queries.push('Bestandsanlage EEG Vergütung Repowering Erweiterung');
  }

  return queries;
}

// MMR-style deduplication: prefer diverse chunks
function mmrDeduplicate(chunks: RetrievedChunk[], maxTokensBudget: number): RetrievedChunk[] {
  const selected: RetrievedChunk[] = [];
  const seen = new Set<string>();
  let totalChars = 0;

  for (const chunk of chunks.sort((a, b) => b.score - a.score)) {
    if (seen.has(chunk.id)) continue;
    // Simple overlap check: skip if another chunk from the same source has nearly same text
    const isDuplicate = selected.some(
      (s) => s.source === chunk.source && levenshteinSimilarity(s.text, chunk.text) > 0.8
    );
    if (isDuplicate) continue;

    totalChars += chunk.text.length;
    if (totalChars > maxTokensBudget * 4) break; // ~4 chars per token estimate

    seen.add(chunk.id);
    selected.push(chunk);
  }

  return selected;
}

function levenshteinSimilarity(a: string, b: string): number {
  const longer = a.length > b.length ? a : b;
  const shorter = a.length > b.length ? b : a;
  if (longer.length === 0) return 1.0;
  const editDistance = editDist(longer, shorter);
  return (longer.length - editDistance) / longer.length;
}

function editDist(a: string, b: string): number {
  const dp: number[][] = Array.from({ length: a.length + 1 }, (_, i) =>
    Array.from({ length: b.length + 1 }, (_, j) => (i === 0 ? j : j === 0 ? i : 0))
  );
  for (let i = 1; i <= a.length; i++) {
    for (let j = 1; j <= b.length; j++) {
      dp[i][j] = a[i - 1] === b[j - 1]
        ? dp[i - 1][j - 1]
        : 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
    }
  }
  return dp[a.length][b.length];
}

export async function retrieveRelevantChunks(input: ProjectInput): Promise<RetrievedChunk[]> {
  const queries = buildQueries(input);
  const allChunks: RetrievedChunk[] = [];

  // Tags to filter by based on user selections
  const tagFilter: string[] = ['pv'];
  if (input.planWallbox || input.hasEV) tagFilter.push('wallbox');
  if (input.planStorage) tagFilter.push('storage');
  if (input.planHeatPump) tagFilter.push('heat_pump');

  try {
    const collection = await getCollection();

    for (const query of queries) {
      const results = await collection.query({
        queryTexts: [query],
        nResults: TOP_K,
        where: tagFilter.length > 1 ? { tags: { $in: tagFilter } } : undefined,
      });

      const ids = results.ids[0] ?? [];
      const distances = results.distances?.[0] ?? [];
      const metadatas = results.metadatas?.[0] ?? [];
      const documents = results.documents?.[0] ?? [];

      ids.forEach((id, idx) => {
        const meta = metadatas[idx] as Record<string, unknown> | null ?? {};
        allChunks.push({
          id: String(id),
          text: String(documents[idx] ?? ''),
          source: String(meta.source ?? ''),
          title: String(meta.title ?? ''),
          status: (meta.status as RetrievedChunk['status']) ?? 'unclear',
          validFrom: String(meta.validFrom ?? ''),
          validUntil: meta.validUntil ? String(meta.validUntil) : null,
          tags: Array.isArray(meta.tags) ? (meta.tags as string[]) : [],
          score: 1 - (distances[idx] ?? 1),
        });
      });
    }
  } catch {
    // Vector DB not available — fall back to static source metadata
    console.warn('Vector DB unavailable, falling back to static sources');
    const staticSources = regulatorySources as RegulatoryDocument[];
    return staticSources
      .filter((s) => tagFilter.some((t) => s.tags.includes(t)))
      .map((s, i) => ({
        id: s.id,
        text: s.chunkText,
        source: s.source,
        title: s.title,
        status: s.status,
        validFrom: s.validFrom,
        validUntil: s.validUntil,
        tags: s.tags,
        score: 1 - i * 0.05,
      }));
  }

  return mmrDeduplicate(allChunks, MAX_CONTEXT_TOKENS);
}
